hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -2205,6 +2276,318 @@ private void listStatusFindKeyInTableCache(
+         *    /a/1...1M files and /a/b/1...1M files.
+         *    /a/1...1M directories and /a/b/1...1M directories.
+         * Listing "/a", will always have the parentID as "a" irrespective of
+         * the startKey value.
+         */
+        // TODO: recursive flag=true will be handled in HDDS-4360 jira.
+        OzoneFileStatus fileStatusInfo = getOzoneFileStatusV1(volumeName,
+                bucketName, startKey, false, null, true);
+
+        if (fileStatusInfo != null) {
+          prefixKeyInDB = fileStatusInfo.getKeyInfo().getParentObjectID();
+          if(fileStatusInfo.isDirectory()){
+            seekDirInDB = metadataManager.getOzonePathKey(prefixKeyInDB,
+                    fileStatusInfo.getKeyInfo().getFileName());
+
+            // Order of seek -> (1) Seek dirs in dirTable. In OM, always the
+            // order of search is, first seek into fileTable and then dirTable.
+            // So, its not required to search again into the fileTable.
+
+            // Seek the given key in dirTable.
+            getDirectories(recursive, startKey, numEntries,
+                    fileStatusList, volumeName, bucketName, seekDirInDB,
+                    prefixKeyInDB, prefixPath, countEntries);
+
+          } else {
+            seekFileInDB = metadataManager.getOzonePathKey(prefixKeyInDB,
+                    fileStatusInfo.getKeyInfo().getFileName());
+            // begins from the first sub-dir under the parent dir
+            seekDirInDB = metadataManager.getOzonePathKey(prefixKeyInDB, "");
+
+            // Seek the given key in key table.
+            countEntries = getFilesFromDirectory(fileStatusList, seekFileInDB,
+                    prefixPath, prefixKeyInDB, startKey, countEntries,
+                    numEntries);
+            // Seek the given key in dir table.
+            getDirectories(recursive, startKey, numEntries, fileStatusList,
+                    volumeName, bucketName, seekDirInDB,
+                    prefixKeyInDB, prefixPath, countEntries);
+          }
+        } else {
+          // No key exists for the given startKey.
+          // TODO: HDDS-4364: startKey can be a non-existed key
+          return Collections.emptyList();
+        }
+      }
+    } finally {
+      metadataManager.getLock().releaseReadLock(BUCKET_LOCK, volumeName,
+              bucketName);
+    }
+
+    // Convert results in cacheKeyMap to List
+    for (OzoneFileStatus fileStatus : fileStatusList) {
+      // refreshPipeline flag check has been removed as part of
+      // https://issues.apache.org/jira/browse/HDDS-3658.
+      // Please refer this jira for more details.
+      refreshPipeline(fileStatus.getKeyInfo());
+
+      // No need to check if a key is deleted or not here, this is handled
+      // when adding entries to cacheKeyMap from DB.
+      if (args.getSortDatanodes()) {
+        sortDatanodeInPipeline(fileStatus.getKeyInfo(), clientAddress);
+      }
+    }
+    fileStatusFinalList.addAll(fileStatusList);
```

Comment:
When startKey is passed, we seek to that and to construct keyName we used keyArgs keyName as keyName and construct the paths.

But let's say

KeyArgs keyName /a/b
startkey is /f/d

Then we have used parentObjectID and seek and then return all keys in directory /f, but we construct keyNames with prefix as "/a/b".

Do you think we need to have a check when startKey is passed is the parentObjectID is same when recursive is false. To avoid this kind of corner case scenarios?

Reply:
As I'm implementing only recursive=false cases in this patch, it will return list with only immediate files/dirs. In your example, the keyname is ` /a/b`, then it will list only upto  `/a/b/f`. It won't have ` /a/b/f/d` entry in the result list.

KeyArgs keyName `/a/b`
startkey will be in the fashion `/a/b/f`. It will be full path, right?.

I believe the case what you mentioned will occur when I'm doing recursive=true in jira HDDS-4360, am I correct?

Reply:
Based on offline discussions, added a case to safe check to verify `startKey` is an immediate child of `keyName`.

Following are the valid parentKey formats:
`    parentKey="" or parentKey="/" or parentKey="/a" or parentKey="a"`
Following are the valid startKey formats, which is called as childKey:
`    childKey="/" or childKey="/a/b" or childKey="a/b"`